### PR TITLE
Header parser impl

### PR DESCRIPTION
**Pull request by oleorhagen**
Implementation of the header-parser, minus the meta-data parsing, which will be handled as a follow-up in https://northerntech.atlassian.net/browse/MEN-6424.

For now it's directly importing the `vendor/expected` lib, but I will move this out into our own wrapper in a separate PR in parallel, and then rebase on top of it.

### CODE REVIEW

File: artifact/v3/header/header.hpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace header {
+
+using namespace std;
+
+using nonstd::nullopt;
+using nonstd::optional;
```
Comment (kacf) [inline artifact/v3/header/header.hpp:37]: We should do like we did for `expected`, and relabel it `mender::common::optional` instead.

File: artifact/v3/header/header.hpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace error = mender::common::error;
+
+//
+// +---header-info
+//
+
+enum class Payload {
+	RootfsImage,
```
Comment (kacf) [inline artifact/v3/header/header.hpp:48]: Do we actually need `RootfsImage`?

In the new implementation there shouldn't be any difference between `ModuleImage` and `RootfsImage`, because the details are handled by the Update Module.

File: artifact/v3/header/header.hpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	EmptyPayload,
+};
+
+using PayloadType = pair<Payload, string>;
+
+// {'artifact_name', 'artifact_group'}
+struct Provides {
+	string artifact_name;
+	optional<string> artifact_group;
+};
```
Comment (kacf) [inline artifact/v3/header/header.hpp:59]: And a map as well, or?

File: artifact/v3/header/token.hpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+#include <unordered_map>
+#include <string>
+#include <regex>
+
+#include <common/log.hpp>
+#include <common/expected.hpp>
+
+const std::regex artifact_script_regexp {
+	"scripts/Artifact(Install|Reboot|Rollback|RollbackReboot|Failure)_(Enter|Leave)_[0-9][0-9](_[^[:space:]]+)?",
```
Comment (kacf) [inline artifact/v3/header/token.hpp:27]: Should use `(_\S+)?` at the end instead (this is what [the original](https://github.com/mendersoftware/mender-artifact/blob/master/artifact/scripter.go#L50) uses).

File: artifact/v3/header/token.hpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+			return Type::ArtifactScripts;
+		}
+		if (std::regex_match(type_name, regex("headers/[0-9]{4}/type-info"))) {
+			return Type::ArtifactHeaderTypeInfo;
+		}
+		if (regex_match(type_name, regex("headers/[0-9]{4}/meta-data"))) {
+			return Type::ArtifactHeaderMetaData;
+		}
+		log::Error("Unrecognized token: " + type_name);
+		return Type::Unrecognized;
```
Comment (kacf) [inline artifact/v3/header/token.hpp:102]: This might not be the right place to handle it (and maybe you already have), but I'll just point it out: I remember that I fixed a bug in mender-artifact a long time ago where it couldn't read an artifact which was artificially put together using `tar`. While this is of course not common, it has its debugging uses and occasionally even in production. Notice the difference between them:

File list from mender-artifact:

```
headers/0000/type-info
```

File list from `tar`:

```
headers
headers/0000
headers/0000/type-info
```

In other words it contains the directories leading up to the file as well. This should be filtered out by looking at file type.

File: common/json.hpp [hunk 6]
```diff
@@ -121,6 +121,16 @@ ExpectedJson Load(string json_str);
 ExpectedJson Load(istream &str);
 ExpectedJson Load(io::Reader &reader);
 
+
+using ExpectedStringVector = expected::ExpectedStringVector;
+using ExpectedKeyValuesMap = expected::expected<unordered_map<string, string>, error::Error>;
+
+// TODO - Templatize
```
Comment (kacf) [inline common/json.hpp:128]: AFAIK you cannot template on return type alone, so this may not be possible.

**Review by kacf — COMMENTED** [review]
> Obviously not completely clean and ready for final review, but I would appreciate a view through still.

Based on this comment I focused on reviewing the headers. The interface looks good to me, I couldn't spot any major issues.

File: artifact/v3/header/header.hpp (hunk 2 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:48]: Very good point!

File: artifact/v3/header/token.hpp (hunk 4 above)
Comment (oleorhagen) [reply artifact/v3/header/token.hpp:27]: It's the same thing, but sure.

Comment (kacf) [reply artifact/v3/header/token.hpp:27]: Oh is it? I thought it was only actual space. Then either is fine.

File: artifact/v3/header/header.hpp (hunk 2 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:50]: I decided to keep it in here, due to the fact that we need to workaround that `type` is not written in `type-info:payload_type`

File: artifact/v3/header/header.hpp (hunk 1 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:39]: I'll create a separate PR for this one, and subsequently rebase on top of it.

File: artifact/v3/header/artifact_scripts.cpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace header {
+namespace artifact_scripts {
+
+ExpectedArtifactScript Parse(io::Reader &reader) {
+	log::Info("TODO - Parse the scripts names - and verify");
+	// Read the whole script to a string, and store it in memory
+	vector<uint8_t> buf(4096); // TODO - Use a stream, and variable size
```
Comment (kacf) [inline artifact/v3/header/artifact_scripts.cpp]: I take it this means that the file storing is not part of this PR?

File: common/json/json.cpp [hunk 8]
```diff
@@ -45,6 +48,56 @@ error::Error MakeError(JsonErrorCode code, const string &msg) {
...
+		return expected::unexpected(
+			MakeError(JsonErrorCode::ParseError, "The JSON object is not an array"));
+	}
+	vector<string> vector_elements {};
+	size_t vector_size {j.GetArraySize().value()};
+	for (size_t i = 0; i < vector_size; ++i) {
+		auto element = j.Get(i).and_then([](const json::Json &j) { return j.GetString(); });
+		if (!element) {
+			return expected::unexpected(MakeError(
+				JsonErrorCode::ParseError, "Unexpected element type in the string array"));
```
Comment (kacf) [inline common/json/json.cpp:63]: These aren't really `ParseError`s though, are they? More like `WrongFormatError` or something like that.

Goes for all of the errors below.

**Review by kacf — COMMENTED** [review]
Posting what I have, I will have to continue on Monday.

Note to self: Continue at `while (tok.type != token::Type::EOFToken)`.

File: artifact/v3/header/header.cpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+		if (!myfile.good()) {
+			// TODO - Return what sort of error (?)
+			return expected::unexpected(parser_error::MakeError(
+				parser_error::Code::ParseError,
+				"Failed to create a file for writing the Artifact script: " + artifact_script_path
+					+ " to the filesystem"));
+		}
+		io::StreamWriter bw {myfile};
+
+		auto err = io::Copy(bw, *tok.value);
```
[the remark this answers is not in the document]
Comment (kacf) [reply artifact/v3/header/header.cpp:90]: I guess you already answered this f2f.

File: artifact/v3/header/artifact_scripts.cpp (hunk 7 above)
Comment (oleorhagen) [reply artifact/v3/header/artifact_scripts.cpp]: Nope, removed.

File: artifact/v3/header/header.cpp [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+			"Failed to parse the header-info: " + expected_info.error().message));
+	}
+	header.info = expected_info.value();
+
+	tok = lexer.Next();
+	vector<ArtifactScript> state_scripts {};
+	while (tok.type == token::Type::ArtifactScripts) {
+		log::Trace("Parsing state script...");
+		const string artifact_script_path =
+			conf.artifact_scripts_filesystem_path + "/" + tok.TypeToString();
```
Comment (vpodzime) [inline artifact/v3/header/header.cpp:78]: Should use `paths::Join()`.

File: artifact/v3/header/header.cpp [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+		ofstream myfile(artifact_script_path);
+		log::Trace("state script name: " + tok.TypeToString());
+		if (!myfile.good()) {
+			// TODO - Return what sort of error (?)
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply artifact/v3/header/header.cpp:82]: I did [this](https://github.com/mendersoftware/mender/blob/feature-c%2B%2B-client/common/log/platform/boost/boost_log.cpp#L200) in the logging code. But if you don't want to use a custom error code, just use the `generic_category` and the `errno`.

File: artifact/v3/header/header.cpp [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+			return expected::unexpected(parser_error::MakeError(
+				parser_error::Code::ParseError,
+				"Failed to create a file for writing the Artifact script: " + artifact_script_path
+					+ " to the filesystem"));
+		}
+		io::StreamWriter bw {myfile};
```
Comment (vpodzime) [inline artifact/v3/header/header.cpp:88]: `bw` is a weird name for this variable. Copy-pasta?

File: artifact/v3/header/header.cpp [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+			// TODO - follow-up (MEN-6204)
+			// return expected::unexpected(expected_meta_data.error());
+			// sub_header.metadata = expected_meta_data.value();
+			// }
+		}
+		log::Trace("sub-header: parsed the meta-data");
+
+		header.subHeaders.push_back(sub_header);
+
+		tok = lexer.Next();
```
Comment (vpodzime) [inline artifact/v3/header/header.cpp:145]: Shouldn't this `Next()` be inside the above `if`? Otherwise this will skip a token if it's not `ArtifactHeaderMetaData`, right?

File: artifact/v3/header/header.hpp [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+struct Provides {
+	string artifact_name;
+	optional<string> artifact_group;
+};
+
+struct Depends {
+	vector<string> device_type;
+	optional<vector<string>> artifact_name;
+	optional<vector<string>> artifact_group;
```
Comment (vpodzime) [inline artifact/v3/header/header.hpp:65]: Shouldn't this be a vector of string pairs?

File: artifact/v3/header/header.hpp [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+// +---header-info
+//
+
+enum class Payload {
+	RootfsImage,
+	ModuleImage,
+	EmptyPayload,
+};
+
+using PayloadType = pair<Payload, string>;
```
Comment (vpodzime) [inline artifact/v3/header/header.hpp:55]: Would a `struct PayloadType` make sense here? What is the string for? I think it would be nice to give the two parts names.

File: artifact/v3/header/header_info.cpp [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+
+using ExpectedPayloadType = expected::expected<vector<PayloadType>, error::Error>;
+
+ExpectedPayloadType ToPayloadTypes(const json::Json &j) {
+	if (!j.IsArray()) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError, "The JSON object is not an array"));
+	}
+	vector<PayloadType> vector_elements {};
+	size_t vector_size {j.GetArraySize().value()};
```
Comment (vpodzime) [inline artifact/v3/header/header_info.cpp:47]: We could initialize the `vector_elements` to the expected size since we know it.

File: artifact/v3/header/header_info.cpp [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+	for (size_t i = 0; i < vector_size; ++i) {
+		auto expected_element =
+			j.Get(i).and_then([](const json::Json &j) { return j.Get("type"); });
+		if (!expected_element) {
+			return expected::unexpected(parser_error::MakeError(
+				parser_error::Code::ParseError,
+				"Failed to get the type from the payload" + expected_element.error().message));
```
Comment (vpodzime) [inline artifact/v3/header/header_info.cpp:54]: Missing colon and a space in the message.

File: artifact/v3/header/header.hpp (hunk 15 above)
Comment (vpodzime) [reply artifact/v3/header/header.hpp:55]: Also, do we even need the string? Isn't it just a string of the type?

File: artifact/v3/header/header_info.cpp [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+
+	auto expected_json = json::Load(reader);
+
+	if (!expected_json) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError,
+			"Failed to parse the header JSON: " + expected_json.error().message));
+	}
+
+	const json::Json header_info_json = expected_json.value();
```
Comment (vpodzime) [inline artifact/v3/header/header_info.cpp:91]: Why not `auto`?

File: artifact/v3/header/header_info.cpp [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+	if (!artifact_name) {
+		return expected::unexpected(
+			parser_error::MakeError(parser_error::Code::ParseError, artifact_name.error().message));
+	}
+	info.provides.artifact_name = artifact_name.value();
+
+	// provides:artifact_group (optional)
+	log::Trace("Parsing the header-info provides:artifact_group (if any)");
+	auto artifact_group = provides.value().Get("artifact_group").and_then(json::ToString);
+	if (!artifact_group && artifact_group.error().code.value() != json::KeyError) {
```
Comment (vpodzime) [inline artifact/v3/header/header_info.cpp:131]: I'd do `Get("artifact_group")` and then `GetString()` if it returns something, but YMMV.

File: artifact/v3/header/header.hpp (hunk 1 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:39]: @kacf actually the `nonstd::optional` wrapper already defaults to `optional` if it exists in `std`.

Should we still wrap it you think?

File: artifact/v3/header/token.hpp [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+	Uninitialized = 0,
+	EOFToken,
+	Unrecognized,
+	HeaderInfo,
+	ArtifactScripts,
+	ArtifactHeaderTypeInfo,
+	ArtifactHeaderMetaData,
+};
+
+unordered_map<const Type, string> type_map {
```
Comment (vpodzime) [inline artifact/v3/header/token.hpp:54]: Could be an array, no need for a hash map.

File: artifact/v3/header/token.hpp [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+		value {make_shared<tar::Entry>(entry)} {
+	}
+
+	const string TypeToString() const {
+		return type_map.at(type);
+	}
+
+private:
+	Type StringToType(const string &type_name) {
+		if (type_name.find("header-info") == 0) {
```
Comment (vpodzime) [inline artifact/v3/header/token.hpp:87]: Shouldn't this be an exact match of the whole string?

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise. Smart approach to extracting real data from the JSON elements!

File: common/json/json.cpp (hunk 8 above)
Comment (oleorhagen) [reply common/json/json.cpp:63]: That makes it easy then! :tada:

File: artifact/v3/header/header_test.cpp [hunk 22]
```diff
@@ -0,0 +1,368 @@
...
+	CreateTestArtifact(tmpdir, "module-image", {"--type test-module-image"});
+
+	std::fstream fs {tmpdir.Path() + "/header.tar"};
+
+	mender::common::io::StreamReader sr {fs};
+
+	ExpectedHeader expected_header = header::Parse(sr);
+
+	ASSERT_TRUE(expected_header) << expected_header.error().message;
+}
```
Comment (kacf) [inline artifact/v3/header/header_test.cpp:368]: Perhaps add test cases for:

* Some files out of order.
* Wrong index of header or payload.

File: artifact/token.hpp [hunk 23]
```diff
@@ -47,6 +52,7 @@ enum class Type {
 
 const unordered_map<const Type, const string> type_map {
 	{Type::Uninitialized, "Uninitialized"},
+	{Type::EOFToken, "EOF"},
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/token.hpp:55]: Ur right- fixing

File: artifact/parser.cpp [hunk 24]
```diff
@@ -108,17 +95,23 @@ ExpectedArtifact Parse(io::Reader &reader) {
...
 			"Got unexpected token " + tok.TypeToString() + " expected 'Header'"));
 	}
-	// auto header = v3::header::Parse(); // TODO (MEN-6178): Enable
+	auto expected_header = v3::header::Parse(*tok.value, config);
+	if (!expected_header) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError,
+			"Failed to parse the header: " + expected_header.error().message));
+	}
+	auto header = expected_header.value();
```
Comment (kacf) [inline artifact/parser.cpp:104]: Are we missing checksumming? (Or am I just missing where it is..?)

File: artifact/v3/header/header.hpp (hunk 1 above)
Comment (kacf) [reply artifact/v3/header/header.hpp:39]: Hmm, in the name of consistency I think I would vote for that. Just to keep "foreign" names out of our cross platform code.

File: artifact/v3/header/header.cpp (hunk 13 above)
Comment (oleorhagen) [reply artifact/v3/header/header.cpp:145]: Absolutely right!!!

Good :eyes:

File: artifact/v3/header/header.hpp (hunk 14 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:65]: Hmm, maybe, but the name is known, so doing it like this the structure of extraction will be:

`header.info.depends.artifact_group -> vector<string>` 

as opposed to:

`header.info.depends["artifact_group"] -> vector<string>`

Where I chose the former, as it's type-checked, and hence typo safe.

File: artifact/v3/header/header.hpp (hunk 15 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:55]: > Isn't it just a string of the type?

No, in fact this needs a value. This is because you can write your own update-modules, and hence the string is free-form.

So we provide the `Type`, (ie, Update-Module), and you as the end user provides the `Update-Module:Typename`.

Does it make sense?

Comment (oleorhagen) [reply artifact/v3/header/header.hpp:55]: We could consider a struct however though.

File: artifact/v3/header/header_info.cpp (hunk 19 above)
Comment (oleorhagen) [reply artifact/v3/header/header_info.cpp:131]: Yeah, could've. I think they both make sense here, so staying with the current impl

File: artifact/v3/header/token.hpp (hunk 21 above)
Comment (oleorhagen) [reply artifact/v3/header/token.hpp:87]: Yup - good find! fixing

File: artifact/v3/header/header.cpp [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+			// auto expected_meta_data = meta_data::Parse(*tok.value);
+			// if (!expected_meta_data) {
+			// TODO - follow-up (MEN-6204)
+			// return expected::unexpected(expected_meta_data.error());
+			// sub_header.metadata = expected_meta_data.value();
+			// }
+		}
+		log::Trace("sub-header: parsed the meta-data");
+
+		header.subHeaders.push_back(sub_header);
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/v3/header/header.cpp:143]: Parts of it, I still don't allow it in the payload `Next` call. But it's kinda disabled by default now. Enabling it is easy, when we want to :)

File: artifact/v3/header/token.hpp (hunk 20 above)
Comment (vpodzime) [reply artifact/v3/header/token.hpp:54]: If you have an enum that starts with `0`, you can have this mapping in an array and use the enum value as an index in the array.

File: common/json/json.cpp [hunk 26]
```diff
@@ -45,6 +48,56 @@ error::Error MakeError(JsonErrorCode code, const string &msg) {
...
+			MakeError(JsonErrorCode::ParseError, "The JSON is not an object"));
+	}
+
+	auto expected_children = j.GetChildren();
+	if (!expected_children) {
+		return expected::unexpected(
+			MakeError(JsonErrorCode::ParseError, expected_children.error().message));
+	}
+
+	unordered_map<string, string> kv_map {};
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply common/json/json.cpp:82]: No, `using ExpectedKeyValueMap = expected::expected<KeyValueMap, error::Error>;`. It can use the new name of the type.

File: artifact/v3/header/header.hpp (hunk 14 above)
Comment (vpodzime) [reply artifact/v3/header/header.hpp:65]: Actually, I meant `artifact_name - artifact_group`  should be pairs, but that's not how it is in the JSON (if I recall correctly what I've seen in the code).

File: artifact/v3/header/header.hpp (hunk 15 above)
Comment (vpodzime) [reply artifact/v3/header/header.hpp:55]: I see, but a `struct` would still be nicer, I think. We do it everywhere else, AFAICT.

File: artifact/parser.cpp (hunk 24 above)
Comment (oleorhagen) [reply artifact/parser.cpp:104]: Oh my, we're not. Fixing.

File: artifact/v3/header/header.hpp (hunk 14 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:65]: Ahh, yeah. Ideally they probably should.

File: artifact/v3/header/header_info.cpp [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+		auto json_element = expected_element.value();
+		if (json_element.IsString()) {
+			string payload_type = json_element.GetString().value();
+			if (payload_type == "") {
+				return expected::unexpected(
+					parser_error::MakeError(parser_error::Code::ParseError, "Empty Payload type"));
+			}
+			Payload type {Payload::RootfsImage};
+			if (payload_type != "rootfs-image") {
+				type = Payload::ModuleImage;
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/v3/header/header_info.cpp:65]: https://github.com/oleorhagen/mender/blob/header-parser-pr/artifact/v3/header/header_info.cpp#L67

Done here.

File: artifact/v3/header/header.hpp (hunk 15 above)
Comment (oleorhagen) [reply artifact/v3/header/header.hpp:55]: Yup, agreed. Moved to a struct.

File: artifact/v3/header/token.hpp [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+private:
+	Type StringToType(const string &type_name) {
+		if (type_name == "header-info") {
+			return Type::HeaderInfo;
+		}
+		if (std::regex_match(type_name, artifact_script_regexp)) {
+			const string script_name = type_name.substr(7 + 1); // Strip the scripts + / prefix
+			type_map[Type::ArtifactScripts] = script_name;
```
Comment (vpodzime) [inline artifact/v3/header/token.hpp:92]: Modifying a global `type_map` from a function in `Token` looks quite fishy.

File: artifact/v3/manifest/manifest.cpp [hunk 29]
```diff
@@ -80,6 +96,14 @@ ExpectedManifest Parse(mender::common::io::Reader &reader) {
 	return m;
 }
 
+string Manifest::Get(const string &key) {
+	auto value = this->map_.find(key);
+	if (value != this->map_.end()) {
```
Comment (vpodzime) [inline artifact/v3/manifest/manifest.cpp:101]: I'd do `if (this->map_.count(key) != 0)`

Comment (oleorhagen) [reply artifact/v3/manifest/manifest.cpp:101]: This is pre-existing. I've only moved it into the `cpp` file.

Also, it's how the `cppreference examples` do it

**Comment by vpodzime** [comment]
@oleorhagen I've rebased this and resolved conflicts (and one `if (err)...` comparison).

File: artifact/v3/header/token.hpp (hunk 28 above)
Comment (kacf) [reply artifact/v3/header/token.hpp:92]: I'd have to agree. `type_map` is a singleton, but `Token` isn't, so this is a dangerous relationship.

File: artifact/v3/header/token.hpp (hunk 5 above)
Comment (oleorhagen) [reply artifact/v3/header/token.hpp:114]: I won't handle this here, and maybe we should discuss this a little more.

So in short this means that we will accept Artifacts which do not align with the format we have described, I know that the regular cli tools create this form of structure, but then we should update our documentation also, no?

Comment (kacf) [reply artifact/v3/header/token.hpp:114]: I wouldn't say that it doesn't align. It's just that the `tar` tool creates a slightly different internal structure compared to the Go tar writer, but it's still semantically valid. It doesn't really change anything, it's just the same thing represented a bit differently.

It should be pretty easy to filter out though, shouldn't it? Something like (pseudo code):

```
if (entry.type == directory) {
    if (entry.path is a prefix of any valid token) {
        ignore();
    } else {
        error("Forbidden entry");
    }
}
```

**Review by kacf — CHANGES_REQUESTED** [review]

File: artifact/v3/header/header.cpp [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+			// sub_header.metadata = expected_meta_data.value();
+			// }
+			tok = lexer.Next();
+		}
+		log::Trace("sub-header: parsed the meta-data");
+
+		header.subHeaders.push_back(sub_header);
+
+		i++;
+		break; // NOTE: We currently do not support multiple payloads
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/v3/header/header.cpp:173]: Hmm, maybe we should keep spinning it actually

File: artifact/v3/header/token.hpp [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+	const string TypeToString() const {
+		return type_map.at(type);
+	}
+
+	const int Index() const {
+		switch (type) {
+		case Type::ArtifactHeaderTypeInfo:
+			// fallthrough
+		case Type::ArtifactHeaderMetaData:
+			return stoi(this->name.substr(8, 4));
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/v3/header/token.hpp:91]: 1. Hmm, ur probably right. I figured it was OK since we've already vetted the string.
2. fixing.

File: artifact/parser_test.cpp [hunk 32]
```diff
@@ -39,6 +40,8 @@ class ParserTestEnv : public testing::Test {
 public:
 protected:
 	static void SetUpTestSuite() {
+		mender::common::log::SetLevel(mender::common::log::LogLevel::Trace);
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/parser_test.cpp:43]: Hmm, I don't agree here actually. I think it should always be enabled. However, it should only be shown on a failing test, which is more up to the test framework I would think.

But if ur adamant...

File: artifact/v3/header/header.cpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/v3/header/header.cpp:173]: Added this:

```
		// NOTE: We currently do not support multiple payloads
		if (i != 0) {
			return expected::unexpected(parser_error::MakeError(
				parser_error::Code::ParseError,
				"Multiple header entries found. Currently only one is supported"));
		}
```
To the start of the loop for now

File: artifact/v3/header/token.hpp (hunk 5 above)
Comment (oleorhagen) [reply artifact/v3/header/token.hpp:114]: Hmm, alright. But keeping it out of this PR

Comment (kacf) [reply artifact/v3/header/token.hpp:114]: Fair enough.

File: artifact/parser_test.cpp (hunk 32 above)
[the remark this answers is not in the document]
Comment (kacf) [reply artifact/parser_test.cpp:43]: Not disagreeing with that! But right now it will be noisy. And even if we want to improve the test framework, it would be better to enable Trace level globally, not here in this single test, or?

**Comment by kacf** [comment]
> [oleorhagen](https://github.com/oleorhagen) requested a review from [kacf](https://github.com/kacf)

Did you push any changes..? I can't see anything new.

File: artifact/parser_test.cpp (hunk 32 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply artifact/parser_test.cpp:43]: If you give the green light, I'm on it asap!

[the remark this answers is not in the document]
Comment (kacf) [reply artifact/parser_test.cpp:43]: As long as it's hidden for passing tests, I'm all for it. I'd recommend a new PR for this though, it's not really related to this code.

**Review by kacf — APPROVED** [review]
🎆 🎇
